specifications/tcp/tcp.tla

```diff
@@ -0,0 +1,330 @@
+    /\ network' = [ network EXCEPT ![local] = Tail(network[local])]
+    /\ connstate' = [connstate EXCEPT ![local] = "CLOSED"]
+
+TimeWait(local, remote) ==
+    /\ local # remote
+    /\ connstate[local] = "TIME-WAIT"
+    /\ tcb[local]
+    /\ UNCHANGED network
+    /\ connstate' = [connstate EXCEPT ![local] = "CLOSED"]
+    /\ tcb' = [tcb EXCEPT ![local] = FALSE]
+    \* TODO: Timeout=2MSL
+
+Note2(local, remote) ==
+    (* Note 2: A transition from FIN-WAIT-1 to TIME-WAIT if a FIN is received
+       and the local FIN is also acknowledged. *)
+    /\ local # remote
+    /\ connstate[local] = "FIN-WAIT-1"
+    /\ UNCHANGED tcb
+    \* The local FIN is also acknowledged, i.e., skipping the CLOSING state.
+    /\ IsPrefix(<< "FIN", "ACKofFIN" >>, network[local])
+    \* ??? The RFC doesn't explictly mention this, but we send the ACKofFIN here.
+    /\ network' = [ network EXCEPT ![remote] = Append(@, "ACKofFIN"),
+                                   ![local] = SubSeq(network[local], 3, Len(network[local]))]
+    /\ connstate' = [connstate EXCEPT ![local] = "TIME-WAIT"]
+
+System ==
+    \E local, remote \in Peers :
+        \/ SynSent(local, remote)
+        \/ SynReceived(local, remote)
+        \/ Listen(local, remote)
+        \/ Established(local, remote)
+        \/ FinWait1(local, remote)
+        \/ FinWait2(local, remote)
+        \/ Closing(local, remote)
+        \/ LastAck(local, remote)
+        \/ TimeWait(local, remote)
+        \/ Note2(local, remote)
+
+Note3(local, remote) ==
+    (* Note 3: A RST can be sent from any state with a corresponding transition 
+       to TIME-WAIT (see [70] for rationale). Similarly, receipt of a RST from
+       any state results in a transition to LISTEN or CLOSED. *)
+    \* ??? "...can be sent from any state with a corresponding transition to TIME-WAIT"
+    \* ??? could also be interpreted as those states from where we can tranisiton to TIME-WAIT,
+    \* ??? i.e., FIN-WAIT-1, FIN-WAIT-2, CLOSING.
+    /\ local # remote
+    /\ UNCHANGED tcb
+    /\ \/ /\ tcb[local] \* ??? RFC 9293 doesn't mention this, but it seems impossible to send a RST without a TCB.
+          /\ network' = [ network EXCEPT ![remote] = Append(@, "RST")]
+          /\ connstate' = [connstate EXCEPT ![local] = "TIME-WAIT"]
+       \/ /\ IsPrefix(<<"RST">>, network[local])
+          /\ network' = [ network EXCEPT ![local] = Tail(network[local])]
+          /\ \/ connstate' = [connstate EXCEPT ![local] = "LISTEN"]
+             \/ connstate' = [connstate EXCEPT ![local] = "CLOSED"]
+
+Reset ==
+    \E local, remote \in Peers :
+        Note3(local, remote)
+
+Next ==
+    \/ System
+    \/ Reset
+    \/ User
+
+Spec ==
+    /\ Init
+    /\ [][Next]_vars
+    /\ WF_vars(System)
+    \* Would get stuck in SYN-SENT if we don't assert a user command.
+    /\ WF_vars(\E local, remote \in Peers: CLOSE_CLOSE_WAIT(local, remote))
+
+-----------------------------------------------------------------------------
+
+Inv ==
```

Comment:
I also experimented with the following liveness properties, but neither of them holds.  They might not be properties, or the current fairness constraint is insufficient (in light of the user commands).

```tla
IsEstablished ==
    \A p \in Peers :
        \* Violated by behaviors that reset or close before a connection is established.
        connstate[p] = "SYN-SENT" ~> connstate \in [ Peers -> {"ESTABLISHED"} ]

IsEstablishedOrTerminated ==
    \* A connection may remain ESTABLISHED indefinitely or it gets closed eventually (an RST may 
    \* non-deterministically go to LISTEN). 
    <>[](connstate \in [ Peers -> {"ESTABLISHED", "CLOSED", "LISTEN"} ] /\ network = [p \in Peers |-> <<>>])
```

I also experimented with a `Discard` action to drop network packets when a peer is `CLOSED`, and to drop `RST` when in `LISTEN`.

```tla
Discard(peer) ==
    \* ??? A closed socket can't receive data.
    \/ /\ connstate[peer] = "CLOSED"
       /\ network' = [ network EXCEPT ![peer] = <<>> ]
    \/ \* ??? A LISTEN socket only accepts SYN packets.
       /\ connstate[peer] = "LISTEN"
       /\ IsPrefix(<<"RST">>, network[peer])
       /\ network' = [ network EXCEPT ![peer] = Tail(network[peer]) ]
    /\ UNCHANGED <<tcb, connstate>>
```

Reply:
Would you have the weaker property
`\A p \in Peers : connstate[p] = "SYN-SENT" ~> connstate[p] \in {"ESTABLISHED", "CLOSED"}` ?